### ISSUE TITLE
Allow early stopping without validation set

### ISSUE THREAD

**Issue by fulldecent** [Microsoft/LightGBM#801]
Currently the early stopping feature requires the user to specify a validation set. Instead, the validation set should be optional. Reasonable behavior should happen if it is unspecified.

> early_stopping_rounds (int) – Activates early stopping. Requires at least one validation data and one metric If there’s more than one, will check all of them Returns the model with (best_iter + early_stopping_rounds) If early stopping occurs, the model will add ‘best_iteration’ field
>
> Source: http://lightgbm.readthedocs.io/en/latest/python/lightgbm.html#training-api

---

# Benefits of specifying a validation set

1.  The customer may demand model performance be tested against a certain set
2.  Avoid overfitting
3.  Reduce validation computation time versus testing against full data set

 \# 2 is a strawman. Overfitting can still happen because information about fitness of the model to the validation set is fed back into the model via early stopping.

---

# Problems with specifying a validation set

1.  If data is limited, you are wasting valuable data to do validation.
2. If data is limited, the user may not know how much to use.
3. The end user probably forgot to shuffle data before slicing out the validation set.

Here is example code I've seen too many times.

    split = round(len(x_train)*0.95)
    x_train, y_train, x_valid, y_valid = x_train[:split], y_train[:split], x_train[split:], y_train[split:]

Who chose 5%, and who says the observations at the end are best to validate with?

---

# Solution

The goal of early stopping is to decide if any of the latest X rounds has improved performance versus a baseline, according to some metric. **This is a hypothesis and we can test it using a sample of the population at a required confidence level!**

* Allow specifying validation set as a portion (0,1]. The training set will be randomly sampled that many times for each validation round.
* Allow specification of a confidence level (0,1], set a default based on the \#stopping rounds. The validation size will be calculated based on the evaluation metric and the desired confidence.

**Comment by Laurae2** [comment]
@fulldecent There are several issues with your solution:

* Any sample going in the train set leaks and therefore you are overfitting immediately
* Validation may require cross-validation, especially with noisy datasets and unbalanced datasets
* Not every metric is created equal, for instance F1-score requires a quadratic improvement to increase the same way a scaled MAE do
* Some datasets lead to unstable improvements (aka Bosch + AUC)
* Even with 100 observations (or even 50) in a dataset, LightGBM can perform exceptionally well like xgboost on new data if the validation scheme was setup properly (this is one of the first stuff to think about before a project starts, anyway)

The reasonable behavior for early stopping is requesting the user to specify a test set. Gradient descent and other optimization algorithms are prone to overfit.

If the user wants to avoid overfitting a single holdout set, he/she can use multiple holdout sets (or cross-validation).

Validating a model is a crucial part of ML, it must be thought before it even happens (e.g at the beginning of the project, not in the middle of it). I've seen too many times validation being performed on the fly at the last minute, where features leak (aka the infamous person ID) inside the train set (and then, the model is literally junk on new real data). In this case, what must be changed is the user / lead, not the program (LightGBM is a tool to help in the process if it was prepared [see H2O Flow], not a tool to create the process from scratch).

> 2 is a strawman. Overfitting can still happen because information about fitness of the model to the validation set is fed back into the model via early stopping.

No, if you use (repeated) cross-validation.

> If data is limited, you are wasting valuable data to do validation.

No, you just can do (repeated) cross-validation and then train a model on all data.

> If data is limited, the user may not know how much to use.

Repeated validation is always possible. Small data would make it even faster.

> The end user probably forgot to shuffle data before slicing out the validation set.

It is the user fault in this case.

> Who chose 5%, and who says the observations at the end are best to validate with?

The user.

---

Also, more specifically:

> Allow specifying validation set as a portion (0,1]. The training set will be randomly sampled that many times for each validation round.

Leakage if not data is sampled during training, regular validation set if sampled before training.

> Allow specification of a confidence level (0,1], set a default based on the #stopping rounds. The validation size will be calculated based on the evaluation metric and the desired confidence.

Formula? (also, how do you handle quadratic metrics? etc.)

**Comment by fulldecent** [comment]
Thank you for the deep explanation. I can see that even k-fold cross validation is not a reasonable default for everything and users must need to manually set up validation for every project.

**Comment by fulldecent** [comment]
Just documenting my end result here. I have found another approach to achieve what I am looking for. 

```
classifier = lgb.Booster(
    params=params, 
    train_set=lgb_train_set,
)

result = lgb.cv(
    init_model=classifier,
    params=params, 
    train_set=lgb_train_set,
    num_boost_round=1000,
    early_stopping_rounds=20,
    verbose_eval=20
)
```
